### DISCUSSION TITLE
How to Pass an Auth Token

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#1378] [Q&A] [answered]
My editor is cloud-based and will run in a browser.  I  need to pass an auth token to my language server, which is running in a web worker, so that I can use it while making a call to an external REST API during scoping.  Think of it like calling `npm install`, except on-the-fly with credentials.   I'm not exactly sure how I should go about passing the token from the browser request to the worker.

My first thought was to hand it in during dependency injection and then access it via shared services.  But that seems like a possible misuse of the DI mechanism.  Is there a recommended way for me to safely pass information like this to my scope provider and then update it when the token needs a refresh?

**Comment by msujew** [comment]
Hey @ballcoach12, good to see you back :)

First off, it's probably best to pass the token during the `initialize` request of the language client. You can use the `initializationOptions` provided by the language client and receive those on the language server side. See [here](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#initialize). Afterwards, create a new DI service that holds the token and registers a callback to `shared.lsp.LanguageServer.onInitialize(params => ...)` in the constructor.

Then during runtime, if the token changes, you can just send a notification to the language server. You can always listen to any custom notification:

```ts
shared.lsp.Connection?.onNotification('newToken', token => {
  ...
});
```

**Reply by ballcoach12** [reply]
Thanks, @msujew! I'm glad to be back on Langium work.  I had to detour for a few weeks to write our package manager, which we are now integrating.

I'll implement your suggestion and report back.

**Reply by ballcoach12** [reply]
@msujew , from what I can tell, initialization of the language server commences for my scenario in `main-browser.ts`, specifically in the call to `startLanguageServer()`.  So I suspect that my code needs to look something like this in order to hook into the initialization process and provide the token, and then expose a function for keeping it up to date:

```
declare const self: DedicatedWorkerGlobalScope;

/* browser specific setup code */
const messageReader = new BrowserMessageReader(self);
const messageWriter = new BrowserMessageWriter(self);
const connection = createConnection(messageReader, messageWriter);

// Inject the shared services and language-specific services
const { shared } = createDtsAndJsonQueryServices({ connection, ...EmptyFileSystem });

shared.lsp.LanguageServer.onInitialize(param =>  {
    param.initializationOptions = {
        token: "somejwttoken",
    }
})
// Start the language server with the shared services
startLanguageServer(shared);

export function setToken(t: string): void {
    connection.sendNotification('newToken', t);
}
```
Do I have that right? If so, then the only other thing I will have to figure out is how to get the cookie that contains the token.

**Reply by msujew** [reply]
Almost right, but you reversed the direction of the messages. You need to read the `initializationOptions` and not write to them. Also, you need to listen to the notification, not send it yourself. The server receives the info, and doesn't send it.

**Reply by ballcoach12** [reply]
I should probably provide more details on the implementation that I have so far.  

I have a DI service that manages the token and looks like this:

```typescript
export interface TokenService {
    getToken(): string;
    setToken(t: string): void;
}

export class DtsTokenService implements TokenService {

    private _services: LangiumServices;
    private token  = "";

    constructor(services: LangiumServices){
        
        this._services = services;
        this._services.shared.lsp.LanguageServer.onInitialize(param => {
            //set the token that was passed to the language server at startup
            this.token = (param.initializationOptions as DtsInitializationOptions).token;
        });
        this._services.shared.lsp.Connection?.onNotification('newToken', t => {
            this.token = t;
        });
    }    

    getToken(): string {
        return this.token;
    }
    setToken(t: string) {
        this.token = t;
    }
}
```

What I seem to be misunderstanding is how to give the token to the language server during initialization.  `main-browser.ts` seems like the most logical place to do that.  I assumed that `LanguageServer.onInitialize()` was  the place where I would pass the token to the server.  But it sounds like that's the callback that occurs after initialization.

**Reply by msujew** [reply]
> What I seem to be misunderstanding is how to give the token to the language server during initialization. main-browser.ts seems like the most logical place to do that.

Once you have the `TokenService` set up, you are done with the language server. The rest needs to be done on the language **client side**. When initializing the language client, you just need to set the `initializationOptions` (see [here for monaco-languageclient](https://github.com/TypeFox/monaco-languageclient/blob/168dce83ab054de88f1c9a8f78b8581da8228632/packages/client/src/monaco-language-client.ts#L15)), and just perform `sendNotification` when you notice that the auth token has changed in the main thread.

**Reply by ballcoach12** [reply]
Ah! OK, now I understand.  I was making it harder than it needed to be.  I just need to spin up a language client in my calling app and use it to pass the token.

**Reply by msujew** [reply]
>  I just need to spin up a language client in my calling app and use it to pass the token.

Wait, what? How are you using Langium as a language server without a language client in place? It at least sounds like you're not using a language client right now?

**Reply by ballcoach12** [reply]
Sorry. I misstated that. I'm looking to use the language client from the monaco-editor-wrapper instance.

**Reply by msujew** [reply]
Ah, I see. Then you can pass the token in the language client config object, see [here](https://github.com/TypeFox/monaco-components/blob/14fa199c35e0ddb7c0f74b808bc5efa805270347/packages/monaco-editor-wrapper/src/languageClientWrapper.ts#L61).

**Comment by ballcoach12** [comment, marked as answer]
For anyone else who has a similar requirement, here is how I implemented this.  In my case, I am using a worker to run the language server in the browser, which means that I am unable to obtain the cookie that contains the token because browser-side code cannot access a cookie.  

My implementation follows these steps:

1. The user logs in and gets a token that is stored in the browser cookie.
2. When the browser loads the editor, it sends a request back to the hosting server to send the token in a header in the response, which can then be accessed by the requesting code in the browser.
3. The browser-side application code gets a language client from the monaco-editor-wrapper and sends a notification to the language server. The token is included in the notification contents.
4. The language server (worker) receives the notification and hands the token off to a custom Token Service that caches it.
5. When my import scope provider needs to make an HTTP request to resolve a package in an imported scope, it obtains the token from the Token Service and places it in a cookie that part of the request that is sent to my package repository.
6. The package repository checks the token and authorizes all requests.
